pkg/operator/nfs/controller.go

```diff
@@ -18,466 +18,393 @@ limitations under the License.
+	r.Log.Info("Reconciling NFSServer StatefulSet", "Operation.Result ", stsop)
+	switch stsop {
+	case controllerutil.OperationResultCreated:
+		r.Recorder.Eventf(cr, corev1.EventTypeNormal, nfsv1alpha1.EventCreated, "%s nfs-server statefulset: %s", strings.Title(string(stsop)), sts.Name)
+		return nil
+	case controllerutil.OperationResultUpdated:
+		r.Recorder.Eventf(cr, corev1.EventTypeNormal, nfsv1alpha1.EventUpdated, "%s nfs-server statefulset: %s", strings.Title(string(stsop)), sts.Name)
+		return nil
+	default:
+		return nil
 	}
 }
 
-func (c *Controller) onUpdate(oldObj, newObj interface{}) {
-	oldNfsServ, ok := oldObj.(*nfsv1alpha1.NFSServer)
-	if !ok {
-		return
-	}
-	oldNfsServ = oldNfsServ.DeepCopy()
-
-	logger.Infof("Received update on NFS server %s in namespace %s. This is currently unsupported.", oldNfsServ.Name, oldNfsServ.Namespace)
-}
-
-func (c *Controller) onDelete(obj interface{}) {
-	cluster, ok := obj.(*nfsv1alpha1.NFSServer)
-	if !ok {
-		return
+func newLabels(cr *nfsv1alpha1.NFSServer) map[string]string {
+	return map[string]string{
+		"app": cr.Name,
 	}
-	cluster = cluster.DeepCopy()
-	logger.Infof("cluster %s deleted from namespace %s", cluster.Name, cluster.Namespace)
 }
 
-func validateNFSServerSpec(spec nfsv1alpha1.NFSServerSpec) error {
-	serverConfig := spec.Exports
-	for _, export := range serverConfig {
-		if err := validateAccessMode(export.Server.AccessMode); err != nil {
-			return err
-		}
-		if err := validateSquashMode(export.Server.Squash); err != nil {
-			return err
-		}
+func newConfigMapForNFSServer(cr *nfsv1alpha1.NFSServer) *v1.ConfigMap {
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Name,
+			Namespace: cr.Namespace,
+			Labels:    newLabels(cr),
+		},
 	}
-	return nil
 }
 
-func validateAccessMode(mode string) error {
-	switch s.ToLower(mode) {
-	case "readonly":
-	case "readwrite":
-	case noneMode:
-	default:
-		return fmt.Errorf("Invalid value (%s) for accessMode, valid values are (ReadOnly, ReadWrite, none)", mode)
+func newServiceForNFSServer(cr *nfsv1alpha1.NFSServer) *v1.Service {
+	return &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Name,
+			Namespace: cr.Namespace,
+			Labels:    newLabels(cr),
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: newLabels(cr),
+			Type:     corev1.ServiceTypeClusterIP,
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "nfs",
+					Port:       int32(nfsPort),
+					TargetPort: intstr.FromInt(int(nfsPort)),
+				},
+				{
+					Name:       "rpc",
+					Port:       int32(rpcPort),
+					TargetPort: intstr.FromInt(int(rpcPort)),
+				},
+			},
+		},
 	}
-	return nil
 }
 
-func validateSquashMode(mode string) error {
-	switch s.ToLower(mode) {
-	case "rootid":
-	case "root":
-	case "all":
-	case noneMode:
-	default:
-		return fmt.Errorf("Invalid value (%s) for squash, valid values are (none, rootId, root, all)", mode)
+func newStatefulSetForNFSServer(cr *nfsv1alpha1.NFSServer) *appsv1.StatefulSet {
```

Comment:
Can we move the functions that generate spec to a new file called spec.go

Reply:
updated, thanks!